subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -1406,12 +1390,175 @@ ssize_t ztls_recvfrom(int sock, void *buf, size_t max_len, int flags,
 
 	if (ret == MBEDTLS_ERR_SSL_WANT_READ ||
 	    ret == MBEDTLS_ERR_SSL_WANT_WRITE) {
-		errno = EAGAIN;
+		ret = -EAGAIN;
+	} else {
+		ret = -EIO;
+	}
+
+	errno = -ret;
+	return -1;
+}
+
+#if defined(CONFIG_NET_SOCKETS_ENABLE_DTLS)
+static ssize_t recvfrom_dtls_client(struct net_context *context, void *buf,
+				    size_t max_len, int flags,
+				    struct sockaddr *src_addr,
+				    socklen_t *addrlen)
+{
+	int ret;
+
+	if (!context->tls->tls_established) {
+		ret = -ENOTCONN;
+		goto error;
+	}
+
+	ret = mbedtls_ssl_read(&context->tls->ssl, buf, max_len);
+	if (ret >= 0) {
+		if (src_addr && addrlen) {
+			dtls_peer_address_get(context, src_addr, addrlen);
+		}
+		return ret;
+	}
+
+	switch (ret) {
+	case MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY:
+		ret = -ECONNRESET;
```

Comment:
What this error code is based on?

Reply:
Well, server closed connection so the socket is not usable anymore. From `recvfrom` man page we have:
```
ECONNRESET
              A connection was forcibly closed by a peer.
```
So I've figure that's the error code to use here.

Reply:
Strange, my recvfrom man page doesn't explicitly mention ECONNRESET (Ubuntu 16.04).

Let's however go thru what's known about ECONNRESET. First of all, "RESET" there unlikely a random word, but kinda tied to TCP's RST packet. A well-known way to get ECONNRESET is when peer closes socket connection on its side without having read all the data already sent by the originator. For example, a client makes an HTTP request, server send a complete page in return, but client reads it partially and closes the connection (e.g. quits in POSIX-like environment). Server will get ECONNRESET in this case. So, it's clear that ECONNRESET is tied with TCP processing. 

Thus, there would be 2 questions:

1. Is ECONNRESET possible with UDP sockets at all?
2. Even if it is, why do you think that "forcibly closed" applies here. Above is given example what is "forcible" or "abrupt" closure is - closing connection without reading all data peer already sent. In case of TLS, a forcible closure is closing the connection on socket level, instead of TLS level. The latter involves sending a closure notification, and that's exactly what you get here.

So, even ignoring first question, what we have here is clean, coordinated closure, not forced at all, so it can't be an error, and should be an EOF return (i.e. len=0) instead.

Reply:
Ok, thanks for elaborating, I'll return 0 then.